### PR TITLE
Add the environment which is used for the release at ilc.desy.de

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Add a `key4hep-desy-release` environment and package configuration
- A few smaller fixes to other packages, mostly adding a missing version or making a dependency more clear

ENDRELEASENOTES

Note: The release that comes out of this environment is currently probably almost identical to the one that is deployed to `sw.hsf.org`.

### CODE REVIEW

File: environments/key4hep-desy-release/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  view: false
+  # More configuration in other config files
+  include:
+  - desy-packages.yaml
+  modules:
+    enable: []
+  concretization: together
+  specs:
+    - key4hep-stack
+    - fccanalyses@commit.df65dc1059691029ac21da51cdbea37d35249681
```
Comment (vvolkl) [inline environments/key4hep-desy-release/spack.yaml:21]: I guess this comes from a `spack add`?
```suggestion
  specs:
    - key4hep-stack
    - fccanalyses@commit.df65dc1059691029ac21da51cdbea37d35249681
```

File: environments/key4hep-desy-release/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+spack:
+  config:
+    install_tree:
+      root: $CVMFS_INSTALL_DIR/spackages
+      projections:
+        all: ${PACKAGE}/${VERSION}/${target}-${os}-${COMPILERNAME}${COMPILERVER}-opt/${HASH}
+        build_type=Debug: ${PACKAGE}/${VERSION}/${target}-${os}-${COMPILERNAME}${COMPILERVER}-dbg/${HASH}
```
Comment (vvolkl) [inline environments/key4hep-desy-release/spack.yaml:7]: For key4hep and the lcg-releases a shorter hash is used, keeps path a bit more readable (but maybe you want to the full hash for continuity with previous releases?)
```suggestion
        all: ${PACKAGE}/${VERSION}/${target}-${os}-${COMPILERNAME}${COMPILERVER}-opt/${HASH:5}
        build_type=Debug: ${PACKAGE}/${VERSION}/${target}-${os}-${COMPILERNAME}${COMPILERVER}-dbg/${HASH:5}
```

File: environments/key4hep-desy-release/desy-packages.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    variants: +lccd
+  dd4hep:
+    variants: +lcio+xercesc+hepmc3+edm4hep
+  podio:
+    variants: +sio
+
+  # "Core external" packages
+  geant4:
+    variants: ~qt+opengl+vecgeom cxxstd=17
+    version: [11.0.1]
```
Comment (vvolkl) [inline environments/key4hep-desy-release/desy-packages.yaml:23]: Note that unlike the old concretizer, the clingo concretizer treats packages.yaml as a "soft preference" and will change versions if there is a conflict for example. To get a hard requirement, add `geant4@11.0.1` to the specs to be installed in the environment and concretize together.

**Review by vvolkl — APPROVED** [review]
Looks good - do you think we can maybe merge desy-packages.yaml and key4hep-common/packages.yaml? I don't see any fundamental differences there.

File: environments/key4hep-desy-release/spack.yaml (hunk 1 above)
Comment (tmadlener) [reply environments/key4hep-desy-release/spack.yaml:21]: In this case I put them here deliberately, mainly to have the exact version for `fccanalyses` under version control. Adding `key4hep-stack` as well, just lets me skip the `spack add` (and also `spack concretize` in the initial setup) and directly do a `spack install` of the environment. Obviously, once there is an existing installation, and the environment should pick up new changes re-concretizing will become necessary again.

File: environments/key4hep-desy-release/desy-packages.yaml (hunk 3 above)
Comment (tmadlener) [reply environments/key4hep-desy-release/desy-packages.yaml:23]: I have been wondering about that actually in some other cases. That is good to know.

File: environments/key4hep-desy-release/spack.yaml (hunk 1 above)
Comment (vvolkl) [reply environments/key4hep-desy-release/spack.yaml:21]: Made a new patch release for fccanalyses: https://github.com/HEP-FCC/FCCAnalyses/releases/tag/v0.3.7

**Comment by tmadlener** [comment]
> do you think we can maybe merge desy-packages.yaml and key4hep-common/packages.yaml? I don't see any fundamental differences there.

There seems to quite some overlap of packages that are configured exactly the same (e.g. `root`, `geant4`, `dd4hep`,...). But then there are also a few that are configured slightly differently (e.g. `marlin +lccd` vs. `marlin ~lccd`, `gdb` also has a few more variants enabled here or `podio +sio` is used here).

I would like some of the configurations that are here, but maybe I can just merge these into the common config? Or would you like to keep that as it is for the time being? Then we would maybe have to split it into a "real" common config and one where we differ.

File: environments/key4hep-desy-release/spack.yaml (hunk 1 above)
Comment (tmadlener) [reply environments/key4hep-desy-release/spack.yaml]: Added that as well to this PR.

**Comment by vvolkl** [comment]
Yes, I meant I wouldn't object if you change the "common" packages.yaml to ressemble what you have in the desy one.

**Review by vvolkl — CHANGES_REQUESTED** [review]
Waiting for merging of `desy-packages.yaml` and `key4hep-common.yaml` before merging the PR.

**Comment by tmadlener** [comment]
Done.

One thing that is not yet entirely clear to me is the differences in the `all` section of the common `packages.yaml`. Specifically in this case I want to have `RelWithDebInfo` for the release on ilc.desy.de, but `packages.yaml` currently has `Release`. I am currently checking if I can simply override this again in the `spack.yaml` of an environment.

**Comment by vvolkl** [comment]
Thanks! Yes, you can add a `packages:` section in the spack.yaml after the `include:` to override settings.

We can probably use RelWithDebInfo too for sw.hsf.org, I think that is worth the bigger binaries.

**Pull request merged**